### PR TITLE
 Temporal: Port `PlainDateTime`'s `from` tests

### PR DESCRIPTION
**Pull request by jessealama**

### CODE REVIEW

**Comment by jessealama** [comment]
Goes along with [this](https://github.com/tc39/proposal-temporal/pull/2142) PR on the proposal-temporal side.

File: test/built-ins/Temporal/PlainDateTime/from/argument-object.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  TypeError,
+  () => Temporal.PlainDateTime.from({}),
+  "empty object throws"
+);
+
+TemporalHelpers.assertPlainDateTime(
+  Temporal.PlainDateTime.from({year: 1976, month: 11, day: 18, millisecond: 123}),
+  1976, 11, "M11", 18, 0, 0, 0, 123, 0, 0,
+  "plain object with month but not month code"
+);
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/argument-object.js:27]: Add one with month code but not month

File: test/built-ins/Temporal/PlainDateTime/from/comma-decimal-separator.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/comma-decimal-separator.js:1]: I'd rename this `argument-string-comma-decimal-separator.js` or `argument-string-comma.js`

File: test/built-ins/Temporal/PlainDateTime/from/leap-second.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+/*---
+esid: sec-temporal.plaindatetime.from
+description: Leap seconds may be accepted or rejected
+features: [Temporal]
+includes: [temporalHelpers.js]
+---*/
+
+TemporalHelpers.assertPlainDateTime(
+  Temporal.PlainDateTime.from("2016-12-31T23:59:60"),
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/leap-second.js:12]: Also test string argument + `{overflow: "reject"}`

File: test/built-ins/Temporal/PlainDateTime/from/minus-sign.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/minus-sign.js:1]: `argument-string-minus.js`?

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js:1]: `argument-object-month.js`?

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.plaindatetime.from
+description: A plain object argument needs to specify a month
+features: [Temporal]
+includes: [temporalHelpers.js]
+---*/
+
+TemporalHelpers.assertPlainDateTime(
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js:11]: I think you have a few similar tests in `argument-object.js` - please take a look

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+assert.throws(
+  TypeError,
+  () => Temporal.PlainDateTime.from({year: 1976, month: undefined, monthCode: undefined, day: 18}),
+  "DateTime.from(required prop undefined) throws"
+);
+
+assert.throws(
+  TypeError,
+  () => Temporal.PlainDateTime.from({year: 1976, day: 18, hour: 15, minute: 23, second: 30, millisecond: 123}),
+  "DateTime.from({ year: 1976, day: 18, hour: 15, minute: 23, second: 30, millisecond: 123 }) throws"
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js:38]:
```suggestion
  "PlainDateTime.from throws with missing month, monthCode"
```

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+assert.throws(
+  RangeError,
+  () => Temporal.PlainDateTime.from({year: 1976, month: 11, monthCode: "M12", day: 18}),
+  "month and monthCode must agree"
+);
+
+assert.throws(
+  TypeError,
+  () => Temporal.PlainDateTime.from({year: 1976, month: undefined, monthCode: undefined, day: 18}),
+  "DateTime.from(required prop undefined) throws"
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js:32]:
```suggestion
  "PlainDateTime.from throws with explicit undefined month, monthCode"
```

File: test/built-ins/Temporal/PlainDateTime/from/overflow-default-constrain.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  Temporal.PlainDateTime.from(date),
+  ...data,
+  "by default, overflow is constrain (overflow options argument absent)"
+);
+
+TemporalHelpers.assertPlainDateTime(
+  Temporal.PlainDateTime.from(date, {overflow: "constrain"}),
+  ...data,
+  "by default, overflow is constrain (overflow options argument present)"
+);
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/overflow-default-constrain.js:24]: Let's also test `{}` and `() => {}` for options

File: test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+/*---
+esid: sec-temporal.plaindatetime.from
+description: Invalid ISO string not acceptable even with overflow = constrain
+features: [Temporal]
+---*/
+
+assert.throws(
+  RangeError,
+  () => Temporal.PlainDateTime.from("2020-13-34T24:60", {overflow: "constrain"}),
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js:12]: Let's test this with all the options from `overflow-default-constrain.js` as well as `"reject"`.

File: test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  "constrain has no effect on invalid ISO string"
+);
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js:14]: Maybe call it `argument-string-out-of-range.js`?

File: test/built-ins/Temporal/PlainDateTime/from/subsecond-precision.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/subsecond-precision.js:1]: `argument-string-subsecond.js`?

File: test/built-ins/Temporal/PlainDateTime/from/time-separators.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/time-separators.js:1]: `argument-string-time-separators.js`?

File: test/built-ins/Temporal/PlainDateTime/from/timezone-ignored.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/Temporal/PlainDateTime/from/timezone-ignored.js:1]: `argument-string-timezone.js`

**Review by Ms2ger — CHANGES_REQUESTED** [review]
I fear I'm a bit repetitive :)

File: test/built-ins/Temporal/PlainDateTime/from/argument-string.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+/*---
+esid: sec-temporal.plaindatetime.from
+description: String arguments are acceptable
+features: [Temporal]
+includes: [temporalHelpers.js]
+---*/
+
+TemporalHelpers.assertPlainDateTime(
+  Temporal.PlainDateTime.from("1976-11-18 15:23:30"),
```
Comment (ptomato) [inline test/built-ins/Temporal/PlainDateTime/from/argument-string.js:12]: I'd use `T` here, you already have tests for alternate time separators elsewhere.

File: test/built-ins/Temporal/PlainDateTime/from/argument-string-malformed.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Comment (ptomato) [inline test/built-ins/Temporal/PlainDateTime/from/argument-string-malformed.js:1]: I'd name this test `argument-string-invalid.js` and set it up similar to this one: https://github.com/tc39/test262/blob/main/test/built-ins/Temporal/ZonedDateTime/prototype/withPlainDate/argument-string-invalid.js
(Even if you don't include that comprehensive of a list of invalid strings at this point, they can be added later)

File: test/built-ins/Temporal/PlainDateTime/from/argument-number.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.plaindatetime.from
+description: A number might be an acceptable argument
```
Comment (ptomato) [inline test/built-ins/Temporal/PlainDateTime/from/argument-number.js:6]: I'd describe this one as "A number is converted to a string, which may be a valid ISO string"

Comment (ptomato) [reply test/built-ins/Temporal/PlainDateTime/from/argument-number.js:6]: or "A number is converted to a string, then to Temporal.PlainDateTime"

File: test/built-ins/Temporal/PlainDateTime/from/argument-object.js (hunk 1 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/argument-object.js:27]: done

File: test/built-ins/Temporal/PlainDateTime/from/comma-decimal-separator.js (hunk 2 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/comma-decimal-separator.js]: sounds good -- done

File: test/built-ins/Temporal/PlainDateTime/from/leap-second.js (hunk 3 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/leap-second.js:12]: done

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js (hunk 5 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js]: sure, done

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js (hunk 6 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js]: you're right -- I removed a couple of redundant tests.

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js (hunk 7 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js]: done

File: test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js (hunk 8 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/month-data-needed.js]: done

File: test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js (hunk 10 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js]: done

File: test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js (hunk 11 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/overflow-invalid-iso-string.js]: sounds good -- done

File: test/built-ins/Temporal/PlainDateTime/from/time-separators.js (hunk 13 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/time-separators.js]: I like it

done

File: test/built-ins/Temporal/PlainDateTime/from/argument-string.js (hunk 15 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/argument-string.js]: good point

done

File: test/built-ins/Temporal/PlainDateTime/from/argument-string-malformed.js (hunk 16 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/argument-string-malformed.js]: I grabbed those strings and tweaked them slightly -- thanks for the suggestion!

File: test/built-ins/Temporal/PlainDateTime/from/argument-number.js (hunk 17 above)
Comment (jessealama) [reply test/built-ins/Temporal/PlainDateTime/from/argument-number.js]: sure, sounds good

done

File: test/built-ins/Temporal/PlainDateTime/from/argument-plaindatetime.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.plaindatetime.from
+description: A Temporal PlainDateTime object is an acceptable argument
```
Comment (ptomato) [inline test/built-ins/Temporal/PlainDateTime/from/argument-plaindatetime.js:6]: There's a similar test, https://github.com/tc39/test262/blob/main/test/built-ins/Temporal/PlainDate/from/argument-plaindate.js that has a description "A PlainDate object is copied, not returned directly" which I think tells a bit more about what we are testing here. I'm basing the corresponding test for Instant on this one, but using that description.

**Review by ptomato — APPROVED** [review]
I can think of a few refinements to some of the tests, but I'd rather not delay this; I'll be writing many of the similar tests for Temporal.Instant, and trying to extend them to other types as well, so I can fix things up as I go along.

**Pull request merged**